Automattic/newspack-newsletters
# feat: constant contact integration

### All Submissions:






### Changes proposed in this Pull Request:

Adds support for Constant Contact. 

To avoid overcomplicating this PR, the branch doesn't include UI for selecting ESP and inputting Constant Contact API key and access token. This will be tackled in another PR immediately after this one lands. Temporarily, API key and access token are set as `wp-config` constants.

The implementation uses the V2 API and requires an API key and Access Token. The V3 API seems to require an oAuth flow which wouldn't be appropriate for this use case since there is no centralized application. If this proves problematic we could rearchitect with a WordPress.com endpoint, but this doesn't seem warranted at the moment. 

Here are a few differences between the Constant Contact and Mailchimp implementations:

- Besides list selection, there is no segmentation support in this implementation. 
- Constant Contact throws an error if the campaign name is a duplicate. To get around this I use a unique ID for the list name. This is a pretty strange and unfriendly approach, and I'd be interested in alternative ideas.
- Constant Contact allows sending to multiple lists, so the lists are presented as a set of `CheckboxControl`s rather than a `SelectControl`. This also requires an additional endpoint to unset a list.
- Constant Contact's campaign object supports seprate`reply_to_email` and `from_email` email addresses. In this implementation the two will always be the same.
- Unlike Mailchimp Constant Contact throws an error if a campaign doesn't have a subject or verified sender email. Because of this, new campaigns are initialized with sender (either the CC account's organization name, the name of the account owner, or a fallback) and the first verified email address. This is actually kind of nice, and might be worth implementing in Mailchimp too. 
- Constant Contact's errors are generally pretty clear, so I did away with the `validate()` function used in the Mailchimp implementation and simply display the errors as returned.

Some resources about the Constant Contact V2:

- Docs: https://developer.constantcontact.com/docs/developer-guides/overview-of-api-endpoints.html
- Tester: https://constantcontact.mashery.com/io-docs
- PHP Library: https://github.com/constantcontact/php-sdk
- PHP Library documentation: http://constantcontact.github.io/php-sdk/

A few questions for consideration about the multiple-ESP architecture in general:

- Should the Test sidebar also be ESP-specific? I had to make a small change so that it uses the correct endpoint, but it is conceivable that other ESPs will require slightly different UI. For example, an ESP that supports sending to only one email at a time.
- The API endpoint registration contains a lot of duplication. Should the endpoint definitions be moved into the parent class `Newspack_Newsletters_Service_Provider_Controller` for endpoints that are universal? cc: @iuravic 
- I opted to keep the Mailchimp and Constant Contact campaign ID meta fields distinct, which conceivably means a customer could switch back and forth between ESPs. This lead me to wonder if the Meta registration of these fields should be moved to the ESP classes?

### How to test the changes in this Pull Request:

You will need a Constant Contact account, an API key, and an Access Token to test.

1. Sign up for a Constant Contact account here: https://www.constantcontact.com/
2. Get an API key here: https://constantcontact.mashery.com/
3. Get an Access token by going [here](https://constantcontact.mashery.com/io-docs), pasting the API key in and clicking Get Access Token.
4. Add API key and Access Token to `wp-config` as follows - the presence of these two keys will cause the plugin to use the Constant Contact implementation:
```
define( 'NEWSPACK_NEWSLETTERS_CONSTANT_CONTACT_API_KEY', 'YOUR-API-KEY' );
define( 'NEWSPACK_NEWSLETTERS_CONSTANT_CONTACT_ACCESS_TOKEN', 'YOUR-ACCESS-TOKEN' );
```
5. Check out the branch and run `composer update` (there is a new dependency). 
6. Test all normal flows: create a newsletter, send tests, change the sender info, change the subject, select and deselect lists, send the campaign. 
7. Try setting invalid API key and Access Token to observe errors. 
8. Verify the appearance of Newsletters is roughly the same in Mailchimp and Constant Contact.

### Other information:
